src/binary_space_partition.rs

```diff
@@ -0,0 +1,449 @@
+use std::{fmt, cmp};
+use std::fmt::Write;
+use rand::distributions::WeightedIndex;
+use rand::prelude::*;
+use rand::rngs::StdRng;
+use rand::Rng;
+use serde::Serialize;
+use crate::error::Result;
+
+struct BspLevel {
+    level: Level,
+    map_subsection_min_size: i32,
+    map_subsection_min_room_width: i32,
+    map_subsection_min_room_height: i32,
+}
+
+
+byond_fn!(fn bsp_generate(width, height, hash, map_subsection_min_size, map_subsection_min_room_width, map_subsection_min_room_height) {
+    match bsp_gen(width, height, hash, map_subsection_min_size, map_subsection_min_room_width, map_subsection_min_room_height) {
+        Ok(s) => Some(s),
+        Err(e) => Some(format!("{e}"))
+    }
+});
+fn bsp_gen(width_as_str: &str,
+    height_as_str: &str,
+    hash_as_str: &str,
+    map_subsection_min_size_as_str: &str,
+    map_subsection_min_room_width_as_str: &str,
+    map_subsection_min_room_height_as_str: &str)
+    -> Result<String>{
+    let default_hash: u64 = rand::thread_rng().gen();
+    let width = width_as_str.parse::<i32>()?;
+    let height = height_as_str.parse::<i32>()?;
+    let mut map_subsection_min_size = map_subsection_min_size_as_str.parse::<i32>()?;
+    let map_subsection_min_room_width = map_subsection_min_room_width_as_str.parse::<i32>()?;
+    let map_subsection_min_room_height = map_subsection_min_room_height_as_str.parse::<i32>()?;
+    if map_subsection_min_size < map_subsection_min_room_width || map_subsection_min_size < map_subsection_min_room_height{
+        map_subsection_min_size = cmp::max(map_subsection_min_room_width, map_subsection_min_room_height) + 1
+    }
```

Comment:
Trying to understand this logic. So if the `map_subsection_min_size` is less than `map_subsection_min_room_width` or `map_subsection_min_room_width` it'll be set to whatever is highest + 1.

From the sounds of it it seems like you want `map_subsection_min_size` to always be atleast 1 more than `map_subsection_min_room_width` and `map_subsection_min_room_height`

If that is the case then I think you might have a bug, you should be using <= not < or else you could end up in the scenario of `map_subsection_min_size` being equal to `map_subsection_min_size`

and you could also simplify this to 

```suggestion
    let map_subsection_min_size = cmp::max(
        map_subsection_min_size_as_str.parse::<i32>()?,
        cmp::max(map_subsection_min_room_width, map_subsection_min_room_height) + 1
    )
```

and avoid using a mutable variable

Reply:
gotcha about the mut, and updated to use your double cmp method over the if then cmp 👍